packages/ngtools/webpack/src/ivy/plugin.ts

```diff
@@ -0,0 +1,486 @@
+      compilation.hooks.finishModules.tap(PLUGIN_NAME, (modules) => {
+        if (compilation.errors.length > 0) {
+          return;
+        }
+
+        const currentUnused = new Set(
+          allProgramFiles.filter((sourceFile) => !sourceFile.endsWith('.d.ts')),
+        );
+        modules.forEach((module) => {
+          const { resource } = module as { resource?: string };
+          const sourceFile = resource && builder.getSourceFile(forwardSlashPath(resource));
+          if (!sourceFile) {
+            return;
+          }
+
+          builder.getAllDependencies(sourceFile).forEach((dep) => currentUnused.delete(dep));
+        });
+        for (const unused of currentUnused) {
+          if (previousUnused && previousUnused.has(unused)) {
+            continue;
+          }
+          compilation.warnings.push(
+            `${unused} is part of the TypeScript compilation but it's unused.\n` +
+              `Add only entry points to the 'files' or 'include' properties in your tsconfig.`,
+          );
+        }
+        previousUnused = currentUnused;
+      });
+
+      // Store file emitter for loader usage
+      compilation[AngularPluginSymbol] = fileEmitter;
+    });
+  }
+
+  private updateAotProgram(
+    compilerOptions: CompilerOptions,
+    rootNames: string[],
+    host: CompilerHost,
+    diagnosticsReporter: DiagnosticsReporter,
+    resourceLoader: WebpackResourceLoader,
+  ) {
+    // Create an ngtsc plugin instance used to initialize a TypeScipt host and program
+    const ngtsc = new NgTscPlugin(compilerOptions);
+
+    const wrappedHost = ngtsc.wrapHost(host, rootNames, compilerOptions);
+
+    // SourceFile versions are required for builder programs.
+    // The wrapped host adds additional files that will not have versions
+    augmentHostWithVersioning(wrappedHost);
+
+    const oldProgram = this.ngtscNextProgram;
+    const program = ts.createProgram({
+      options: compilerOptions,
+      oldProgram,
+      rootNames: wrappedHost.inputFiles,
+      host: wrappedHost,
+    });
+
+    // The `ignoreForEmit` return value can be safely ignored when emitting. Only files
+    // that will be bundled (requested by Webpack) will be emitted. Combined with TypeScript's
+    // eliding of type only imports, this will cause type only files to be automatically ignored.
+    // Internal Angular type check files are also not resolvable by the bundler. Even if they
+    // were somehow errantly imported, the bundler would error before an emit was attempted.
+    // Diagnostics are still collected for all files which requires using `ignoreForDiagnostics`.
+    const { ignoreForDiagnostics, ignoreForEmit } = ngtsc.setupCompilation(program, oldProgram);
+
+    const builder = ts.createEmitAndSemanticDiagnosticsBuilderProgram(
+      program,
+      wrappedHost,
+      this.builder,
+    );
+
+    // Save for next rebuild
+    if (this.watchMode) {
+      this.ngtscNextProgram = ngtsc.getNextProgram();
```

Comment:
Requesting the next program here might be premature, as requesting type check diagnostics (`NgtscPlugin.getDiagnostics`) will create a new `ts.Program` which provides a new next program to use for the next incremental build. I verified that calling `NgtscPlugin.getNextProgram` just before calling `this.createFileEmitter` in the promise handler does indeed return a different `ts.Program`.

Reply:
This was originally later in the process.  As an experimental with the release of TS 4.0, I moved it to its current location and it didn't appear to cause a change in behavior (no crashes, diagnostics appear to be the same, etc.).  TS 4.0 removed an unnecessary assert statement that would previously cause an exception in this case (https://github.com/microsoft/TypeScript/blob/v3.9.2/src/compiler/program.ts#L1239).

Reply:
Yeah I also didn't observe any difference in behavior. I was expecting it to crash due to that assertion, but if it's been removed then surely it does not. I'm not sure if this causes extra memory to be retained, and what the size of a reused `ts.Program` would be. Maybe @alxhub has some more insights here.